### ISSUE TITLE
Biblatex writes `\abx@aux@refcontextdefaultsdone` to the main.aux and triggers a redundant run of `latex` in clean directory

### ISSUE THREAD

**Issue by thebluepotato** [plk/biblatex#1167]
Hi! Back again with my quest of shorter typeset times, similarly to #1165.

A basic MWEB with `\cite`:
```latex
\documentclass{article}
\usepackage{biblatex}

\addbibresource{\jobname.bib}

\begin{filecontents}{\jobname.bib}
@book{key,
  author = {Author, A.},
  year = {2001},
  title = {Title},
  publisher = {Publisher}
}
\end{filecontents}

\begin{document}

This is a test of using cite \cite[foo][bar]{key}.

\printbibliography

\end{document}
```
Causes the following runs with `latexmk -pdf` (and hence Overleaf in all likelihood):
```
pdflatex (1) (initial run)
biber
pdflatex (2) (writes checksum and refcontexts to main.aux)
pdflatex (3) (final PDF, main.aux written to by biblatex)
pdflatex (4) (merely timestamp changes)
```

However, a diff of the PDF between the 3rd and 4th run show that merely timestamps have changed. Even the logs are identical (save for time and number of words). Hence, that 4th run (or 3rd after `biber`) is redundant.

What's ahhepning? The fourth `pdflatex` run is caused by main.aux being written to in the third run by biblatex, adding this line:
```latex
\abx@aux@refcontextdefaultsdone
```
This line expands to:
https://github.com/plk/biblatex/blob/9163c931822ffef222d8c533610edabea651c921/tex/latex/biblatex/biblatex.sty#L9868-L9869

However, the 2nd run (or first after `biber`) already writes refcontexts in main.aux:
```latex
\abx@aux@defaultrefcontext{0}{key}{nty/global//global/global}
```

Which in turn is defined as:
https://github.com/plk/biblatex/blob/9163c931822ffef222d8c533610edabea651c921/tex/latex/biblatex/biblatex.sty#L9865-L9867

Both commands written to the aux file set `refcontextdefaults` to false. And based on the code, when `\abx@aux@defaultrefcontext` is read in the 3rd run, `biblatex` sees the toggle is false and then writes `\abx@aux@refcontextdefaultsdone` to the aux file even though it does nothing more than `\abx@aux@defaultrefcontext` already does and merely "confirms" that the toggle indeed should be false.
https://github.com/plk/biblatex/blob/9163c931822ffef222d8c533610edabea651c921/tex/latex/biblatex/biblatex.sty#L9875-L9879

**TL/DR;** what's the point of writing `\abx@aux@refcontextdefaultsdone` to the aux file? It's only written when `refcontextdefaults` is already false and all it does is set it to false again. `biblatex` already commits the refcontext defaults to the aux file as soon as `blx@bbldone` is true and they set `refcontextdefaults` to false on future runs.
Again, the PDF output is identical save for timestamps between the 3rd and 4th run, meaning we could avoid it (and shave off nearly 25% of compile time for large projects).
So I'm thinking we could either remove `\abx@aux@refcontextdefaultsdone` completely or have it written at the same time as the other `\abx@aux@defaultrefcontext`.

PS: Might be related to #1116?
PPS: of course if we save the BBL file between runs and clean out all the rest, which `latexmk -gg` usually does depending on settings, `blx@bbldone` is already true in the first run, meaning we write the refcontexts one run earlier and write `\abx@aux@refcontextdefaultsdone` in the second latex run already. Still, seems redundant.

**Comment by moewew** [comment]
Some digging suggests that thanks to https://github.com/plk/biblatex/commit/163849de21be859ffc3aed17c843ace7a0f18bd4 (which was added to fix https://github.com/plk/biblatex/issues/396 a couple of days after `\abx@aux@refcontextdefaultsdone` was implemented in https://github.com/plk/biblatex/commit/73e58fcadba58b93759d6b36af9c4c27421d970d) `\abx@aux@refcontextdefaultsdone` indeed no longer does anything useful.

See https://github.com/plk/biblatex/pull/1168.

Now that I think about it, I'm wondering if we need the bool `refcontextdefaults` at all, or whether it could be replaced by something more sophisticated.

**Comment by moewew** [comment]
Yeah, we should be able to get rid of `refcontextdefaults` as well: https://github.com/plk/biblatex/pull/1168/commits/c252d4ed9638dc0d593048dca6109f78ca126353.

**Comment by thebluepotato** [comment]
Optimising 5 years down the line, not bad! Glad you could confirm it doesn't do anything useful 👍🏻

**Comment by moewew** [comment]
Turns out the rerun check in https://github.com/plk/biblatex/commit/c252d4ed9638dc0d593048dca6109f78ca126353 did not do the correct thing if an entry appears in several biblists. That should be fixed with a more robust rerun test in https://github.com/plk/biblatex/commit/fed9745fb5c81134a0b89bb98e8be73c580a3bc8.

**Comment by thebluepotato** [comment]
In a related matter, the `main.run.xml` file also gets updated after the third `latex` run (one of the requests is set to inactive).
Since this seems less crucial than an .aux file, I'm trying to configure `latexmk` to ignore changes to this file. Can I safely do this? Which part of biblatex/Buber uses it and for what purpose?

**Comment by moewew** [comment]
The `*.run.xml` file, which is written by the `logreq` package (https://ctan.org/pkg/logreq), was an attempt to make the several rerun requests (rerun LaTeX, rerun Biber etc.) machine readable. I don't think the project really gained traction, I certainly haven't seen this package used much apart from `biblatex` and I seriously doubt many build tools really analyse its contents (but I don't know what exactly `latexmk` does with that file).

**Comment by moewew** [comment]
I've had a slightly closer look at the `*.run.xml` file and it helps to think of the file as machine-readable version of the `.log`. As such the aim for tools like `latexmk` should not be to run as many compilation steps as necessary to make the `*.run.xml` stable, the aim should be to run as many compilation steps as necessary to make the `*.run.xml` no longer request reruns. This is in contrast to the output and files like `.aux` and `.bbl`, which you want to reach a stable state. Assuming that everything is OK, once you reach a stable state, the `*.run.xml` will no longer request reruns and will eventually settle in a stable state, but this takes one step more.

Take an MWE like the one you posted above. Such a document needs to be compiled with LaTeX, Biber, LaTeX, LaTeX to reach a stable state. The compilation cycle with an additional LaTeX run at the end goes something like this

| Step | Program | `.log`  after the run                  | `.run.xml` after the run         | Comment |
|------|---------|------------------------|------------------|-------|
| 1    | LaTeX   | Run Biber, Rerun LaTeX | biber=1, latex=1 |  No bib data, we need a Biber run. |
| 2    | Biber   |      "                  |   "               |  No change to `.log` or `.xml.run`.  Biber only writes to `.bbl` and `.blg`. |
| 3    | LaTeX   | Rerun LaTeX            | biber=0, latex=1 | Need one further LaTeX run to resolve cross-references and align all internal trackers. |
| 4    | LaTeX   |                        | biber=0, latex=0 | PDF has reached stable state. |
| 5    | LaTeX   |                        | biber=0, latex=0 | No change in output. |

If we treat the `.run.xml` like the `.aux` file, whose contents are not only written, but also consumed and used again by LaTeX, then there is a difference in starting conditions between step 4 (step 3 left us with `biber=0, latex=1`) and step 5 (step 4 gave us `biber=0, latex=0`) and so running again is justified. After step 5 the starting conditions are the same as before, so non further runs are requested.

If we interpret the `.run.xml` file like the `.log` file, it is clear that it does not suggest a further LaTeX run after step 4 (which doesn't mean there should not be a further run, mind you).

So I think if you want `latexmk` to be 'better' about this, then you'd have to ask its author to handle the `.run.xml` differently from 'input' files like `.aux`.

**Comment by thebluepotato** [comment]
`latexmk` reads and interprets the log as well and keeps track of all files it considers as "sources". Hence, whenever an `.aux` file changes, it will trigger a rerun. For some reason, it considers the `.run.xml` file as a source as well and hence any modification to it triggers a rerun. In short, it's exactly like you wrote above and `latexmk` does not interpret the `.run.xml` file at all.
For other users stumbling on this issue and before 3.17 comes out, here's a hotfix to ignore both `\abx@aux@refcontextdefaultsdone` in the `.aux`file as well as the `active` status in `.run.xml` file (I was somehow unable to ignore all of `.run.xml` due to the double extension, so I had to have a narrow rule applying to all XML files). You should put this in your `.latexmkrc` file:
```perl
# Ignore the active status of requests in the *.run.xml file generated by biblatex to avoid an extra run of (lua)latex
$hash_calc_ignore_pattern{'xml'} = '^.+active="(?:0|1)"';

# Temp exclusion of a redundant main.aux write, fixed in biblatex 3.17
$hash_calc_ignore_pattern{'aux'} = 'abx@aux@refcontextdefaultsdone';
```
I've tested the `dev` version of `biblatex` and `biber`, with the fix to ignore the status of the `.run.xml` file and I'm happy to report that there are only two `latex` runs after `biber` 👍🏻
